Mgv7p development

paramat opened the pull request:
![screenshot_20180628_163432](https://user-images.githubusercontent.com/3686677/42044718-9a1eb18a-7af1-11e8-9144-9a63f335363e.png)

This generates a little faster than mgv6.

paramat commented on worlds/World 2/world.mt:
```diff
@@ -1 +1,4 @@
-gameid = default
\ No newline at end of file
+gameid = default
```
These changes to 'World 1-4' are strange, i don't know how they happened, they somehow happened due to me testing mgv7p. I will of course remove these changes before any merge.

paramat commented:
@MoNTE48 I will make a separate PR to add the missing nodes that are needed for mgv7p/mgflat biomes, i will also re-organise default/nodes.lua into sections as it is a mess.

I will try to benchmark generating a larger area.
The load time of the game will be unchanged as core mapgen does not add content to the subgame.

paramat commented:
Updated.
Removed changes to 'World 1-4'.

Bedrock is now placed by this mapgen at y = -64 and below, so added 'mapgen_bedrock' to the subgame. So in this mapgen we will not have to place bedrock using the slower ore method, and bedrock is now continuous below y = -64 instead of a thin layer.

Caves do not generate deep into bedrock, but because they are generated to fill each mapchunk they will penetrate into bedrock by up to 48 nodes.

![screenshot_20170801_214842](https://user-images.githubusercontent.com/3686677/28847264-7fc4e1be-7706-11e7-8226-20bfe242a25a.png)

![screenshot_20170801_215025](https://user-images.githubusercontent.com/3686677/28847267-84366fec-7706-11e7-97e6-9634ce0cfe8d.png)

^ For desktop, chunk size is 80 nodes, and caves reach down to y = -112.

paramat commented:
Font path fix moved to new PR #96

paramat commented:
Rebased on master.
Added code to print to the terminal in-game the average generation time for mapchunks generated in the game session.
Flying over terrain in a quick test i get these results:
Mgv6 61ms
Mgv7p 54ms
I will benchmark the average over a 1000x1000 area.

paramat commented:
Average mapchunk generation time for (-500, -112, -500) (500, 127, 500)
Mgv6 61ms
Mgv7p ms

MoNTE48 commented:
These are excellent results. :+1: 
1) First let's add a full float mapgen. A separate commit.
2) Before continuing the work on v7p, let's fix subgame (?) For the full generation of trees and ornaments. So we can get 100% accurate results.

paramat commented:
> First let's add a full float mapgen. A separate commit.

I don't understand, do you mean a 'flat mapgen'? We already have mgflat.
I could add a new flat mapgen that places bedrock below y = -64 like mgv7p does?

>  let's fix subgame (?) For the full generation of trees and ornaments

Ok will do this next.

MoNTE48 commented:
It is better to make changes to the existing float mapgen.
I do not want to have 100500+ mapgens :)

paramat commented:
Ok, but i think you mean 'flat mapgen' not 'float mapgen' https://github.com/MultiCraftProject/MultiCraft/blob/master/src/mapgen_flat.cpp
I'm happy to customise mgflat with the bedrock changes.

MoNTE48 commented:
@paramat oops,sorry. Yes, i mean flat!

paramat commented:
I updated mgv7p, now mapchunks below 'bedrock level' have a separate and much simpler generation function that places bedrock and does nothing else. This makes mapgen slightly faster.

Then i temporarily added the missing nodes to default_game and ran the benchmark again
Average mapchunk generation time for (-500, -112, -500) (500, 127, 500)
Mgv7p 55ms
So slightly faster than before, even with the extra decorations.

MoNTE48 commented:
@paramat, wow, +1)

MoNTE48 commented:
Wish: can the mountains be lower? Now the mountains in v7 are much higher than in v6, and this makes discomfort on mobile devices.

paramat commented:
I agree they are too steep, i will work on that. But i feel they should remain much higher than mgv6, because mgv6 is boring because it never goes above y = 47.
MT mgv7 mountains go up to y = 200-400, so i will try to make them a little lower in mgv7p and less steep.

paramat commented:
In minetest.conf:
`mgv7_np_mount_height = 256, 112, (1000, 1000, 1000), 72449, 3, 0.6, 2.0`
This controls maximum mountain height in nodes.
The first number (256) controls the average, the second number (112) controls the variation either side of the average. So i will tune these.

MoNTE48 commented:
@paramat I raised the height of the clouds to increase performance (although, perhaps, it does not affect the FPS in any way.) I'm not sure).
Yes, climbing the mountain above the clouds will be very cool! But the height of 200-400 blocks is a lot!
A possible solution is clouds at an altitude of 164 and mountains: 100-200 blocks in height! High mountains should not be very frequent. With a mobile phone is not very convenient to travel around the world and bypass the high mountains constantly.

---
How about large fields (flat space without mountains and pits, only trees and decorations)? Are they less frequent or more frequent than in v6?

paramat commented:
Raising the clouds does not affect performance at all.
I will make the mountains half height, this will also reduce the steepness.

> How about large fields (flat space without mountains and pits, only trees and decorations)? Are they less frequent or more frequent than in v6?

mgv6 has everything on a smaller scale, so mgv7(p) has larger open spaces without cliffs and mountains.

paramat commented:
I have made mountains half height for you to test.
Remember mountain height varies from 0 to 256 nodes over a scale of 1000 nodes so you need to travel a long way to see the range of heights.

paramat commented:
I recommend changing cloud height to the MT default of 120, so you can have mountains rising above the clouds.

MoNTE48 commented:
+1, do it!)

paramat commented:
Ok.

paramat commented:
Soon i'll start work on changing the 'flat' mapgen (a separate PR) to place bedrock below y = -64 as mgv7p does.

paramat commented:
@MoNTE48 are you happy to merge this?

paramat commented:
I'm happy to advise on, improve or bugfix any code i've added in PRs, but apart from that i'm not interested in coding for Multicraft.

MoNTE48 commented:
@paramat Hey. I came back again and would be glad to continue working! I apologize for the long absence

paramat commented:
No problem, i was concerned about you.

Currently this PR has extra code for timing generation, this of course needs removing before merge, shall i remove it now or do you want to see the generation times first?
The good news is it is faster than mgv6.

MoNTE48 commented:
@paramat I think this code can be left for debug build. I plan to do a merge of 0.4.16 minetest_game.
And then include mapgen v7p in the world creation menu.

paramat commented:
The code that times chunk generation should not be in debug build, i will remove it from this pull request soon, then this pull request will be ready for merge.

MoNTE48 commented:
Good. I'm finishing upgrade of MultiCraft subgame to the latest version of minetest_game. I will be happy if you help to make sure that there will be no conflicts and "missing" blocks!

paramat commented:
Yes i can help out a little with the game. Contact me if you need advice.

paramat commented:
Sorry for the delay, will now prepare this for merge.

MoNTE48 commented:
@paramat you can remove TimeTaker in mapgen_v6?

paramat commented:
Updated.
Yes i have now removed timetakers in mgv6 and mgv7p.
This is now ready for review. No longer WIP.
Let me know if you want anything changed :)

MoNTE48 commented:
Excellent! Probably the last thing I would like to change - in mapgen v7 (and v7p) there are a lot of high mountains. Is it possible to make them lower? Mobile players do not like climbing high mountains.

paramat commented:
I have already made mapgen v7p mountains half height on your request.
In mgv7p the maximum mountain height varies from is 128 - 56 to 128 + 56 nodes:
```
+MapgenV7PParams::MapgenV7PParams()
+{
+	np_mount_height      = NoiseParams(128,  56,   v3f(1000, 1000, 1000), 72449, 3, 0.6,  2.0);
```
These mountains are quite rare so players don't have to climb them.
Please test mgv7p and let me know the average value and variation you want for the maximum mountain height.
You can set this in .conf using this line:
`mgv7p_np_mount_height = 128, 56, (1000, 1000, 1000), 72449, 3, 0.6, 2.0`
And adjust the numbers '128' and '56'.
////////////////////

Mgv7 mountains are still very high yes, you can change that yourself by editing this line:
https://github.com/MultiCraftProject/MultiCraft/blob/7577cf848b5e812f1d7959b87979ec5e65f61694/src/mapgen_v7.cpp#L140
'256' is the average maximum height.
'112' is the variation either side of the average.
So the maximum mountain height varies from 256 - 112 to 256 + 112.
///////////////////

Maybe you want the cliffs lower too? I have not lowered those in mgv7p.

paramat commented:
I'll lower the cliffs to half height too, because they are now too high relative to the mountains.

MoNTE48 commented:
@paramat okay. And please, reply on me email.

paramat commented:
![screenshot_20180628_163432](https://user-images.githubusercontent.com/3686677/42044718-9a1eb18a-7af1-11e8-9144-9a63f335363e.png)

I made cliffs half height. The screenshot shows the highest i cliffs i can find. The cliffs are slightly lower than mgv6 cliffs.

MoNTE48 commented:
excellent. By the way, why so much stone? Is it possible to remove a stone biome?

paramat commented:
Hang on, small bugfix coming.

paramat commented:
I fixed an error in noise parameter 'np_mountain' that was distorting the mountain shapes.
Here is a map of 2000x2000 nodes (biomes removed for clarity):

![mc5](https://user-images.githubusercontent.com/3686677/42048464-4586106e-7afb-11e8-90a0-019d51fc289b.png)

paramat commented:
If you have a biome with unwanted stone surface that is either an error in biome registration or a missing node. My screenshots show the old game in the engine repo, not the new game.

MoNTE48 commented:
I understood. We can make a temporary correction, with the addition of an alias, so that there are no errors in registering the biomes? While I'm working I need a new version.

paramat commented:
No need for an alias, just add the missing node or fix the biome registration in default/mapgen.lua.

MoNTE48 commented:
@paramat you can look on my mapgen.lua? And write list missing node and broken biome registrations?
Thank's

paramat commented:
Ok.

paramat commented:
> you can look on my mapgen.lua? And write list missing node and broken biome registrations?

@MoNTE48 which game? Game in engine or game in private repo?

MoNTE48 commented:
@paramat (i'm sorry) Now game in engine

paramat commented:
No problem.

This is easy for you to do. Look at lines 372 to 887 here:
https://github.com/MultiCraftProject/MultiCraft/blob/968954a3cc8dbdb818813ecd9c8476ff94447f7d/games/default/files/default/mapgen.lua#L372
Check that the nodes specified in each biome are present in nodes.lua.

paramat commented:
No problem.

This is easy for you to do, so i can't justify doing this myself.
Look at lines 372 to 887 here:
https://github.com/MultiCraftProject/MultiCraft/blob/968954a3cc8dbdb818813ecd9c8476ff94447f7d/games/default/files/default/mapgen.lua#L372
Check that the nodes specified in each biome are present in nodes.lua.

paramat commented:
Ok, it seems you find that difficult to do :) so i'll look at mapgen.lua for any mistakes.

MoNTE48 commented:
Now I'm a little busy fixing the Android build. Perhaps this is not visible, because the actions take place in another repository. But you are right, I would be grateful, if you look at mapgen.lua

paramat commented:
See #121

paramat commented:
Please remove the WIP label, it's ready for merge.
Please can you merge this? I can always adjust the mapgen later if needed.
Having this PR open is beginning to get in the way of other mapgen work.
